### DISCUSSION TITLE
Program-controlled .panel-tabset

### DISCUSSION THREAD

**Discussion by harrelfe** [quarto-dev/quarto-cli#796] [Feature Requests]
A wishlist item: I find it very convenient to have a chunk produce multiple plots and to have them in tabs.  It would be helpful to have an automatic way to do this from within a single R chunk, e.g.

::: {.panel-tabset .column-page}
```{r}
#| tab: Scatterplot
plot(x, y)
#| tab: Histogram
hist(x, nclass=100)
```
:::

Even better would be to allow some tabs to have non-graphical output (generated html or plain text).

**Comment by jjallaire** [comment]
For more complex composition of cell output into markdown we recommend using `output: asis` (to prevent the normal div structure from being written around output). This should implement what you are looking for:

````markdown
::: {.panel-tabset .column-page}
## Scatterplot

```{r}
#| echo: false
#| output: asis
plot(cars)
```

## Histogram

```{r}
#| echo: false
#| output: asis
hist(rnorm(40), nclass = 5)
```
:::
````

**Reply by harrelfe** [reply]
That's cool.  What I'd really like is a "one chunk" solution.  I can do that with cat('## Histogram\n') etc. within the 'asis' chunk but I'd like a more elegant solution.  In some cases it would be good to be even more dynamic.  If a special `tab` function could be written I'd like to call `tab(..., label)` within a loop where labels are computed.

This was my programmatic attempt, but `read_chunk` did not actually execute the generated R code.
```
maketabs <- function(x, labels=names(x)) {
  .xxx. <<- x
  w <- c('\n::: {.panel-tabset}', '')
  for(i in 1 : length(x)) {
    w <- c(w, paste0('## ', labels[i]), '')
    w <- c(w, paste0('```{r}', paste0('.xxx.[[', i, ']]'), '```', '')
  }
  w <- c(w, ':::', '')
  knitr::read_chunk(lines=w)
}
```

**Comment by jjallaire** [comment]
Okay, we'll be on the lookout for scenarios like this where a common idiom can be easily expressed in yaml. I've been conservative at the outset but I think we'll start to see patterns like this one that make sense to automate.

**Reply by harrelfe** [reply]
I did get this function to work: https://github.com/harrelfe/rscripts/blob/master/maketabs.r

Examples are at https://fharrell.com/post/rflow

Having some `yaml` would be more elegant but the `maketabs` function makes it easy to do something in a loop.  One of the examples above uses a mixture of `html` and `htmlwidgets` in the tabs.

**Comment by mcanouil** [comment]
Closing as completed with the code cell option `panel`, see, for example in `knitr` as engine, <https://quarto.org/docs/reference/cells/cells-knitr.html#cell-output>.

**Reply by jan-abel-inwt** [reply]
Hi there,

thank you for your work! Could you please give a short example for the `panel: tabset`option? 
Which quarto Version do I need? `1.3.433` ?

**Reply by mcanouil** [reply]
````qmd
---
title: "Quarto Playground"
format: html
---

```{r}
#| panel: tabset
#| output: asis
list_of_tables <- rep(list(head(mtcars)), 3)
names(list_of_tables) <- c("Table 1", "Table 2", "Table 3")
for (itable in names(list_of_tables)) {
  cat(sprintf("## %s\n", itable))
  print(knitr::kable(list_of_tables[[itable]]))
}
```
````

**Reply by jan-abel-inwt** [reply]
Thanks a lot!

I tried an example with some plotting and this works fine :)

````
```{r}
#| panel: tabset
#| output: asis
x <- seq(-pi, pi, length.out = 100)

y_sin <- sin(x)
y_cos <- cos(x)

# Create data frames
df_sin <- data.frame(x, y_sin)
df_cos <- data.frame(x, y_cos)

# Use ggplot2 to create the plots as objects
library(ggplot2)
sin_plot <- ggplot(df_sin, aes(x=x, y=y_sin)) + geom_line() + labs(title="Sine Wave", x="x", y="y")
cos_plot <- ggplot(df_cos, aes(x=x, y=y_cos)) + geom_line() + labs(title="Cosine Wave", x="x", y="y")

# Save the plots in a list
list_of_plots <- list(sin = sin_plot, cos = cos_plot)

for(iplot in names(list_of_plots)){
  cat(sprintf("## %s\n", iplot))
  print(list_of_plots[[iplot]])
  cat(sprintf("\n\n"))
}
```
````

Now, I wanted to try the same thing in Python with the Jupyter engine. But I'm already failing at generating empty tabs

````
---
title: "Quarto Playground"
format: html
engine: jupyter
---

## Plots

```{python}
#| panel: tabset
#| output: asis

tabs = {"first", "second"}

for tab in tabs:
    print(f"## {tab} \n \n")
```
````

what have i done wrong?

**Reply by cscheid** [reply]
The combination of `panel: tabset` and `output: asis` is unlikely to work well unless you know exactly what you're doing. If you choose to go down that route, I recommend that you first start inspecting the markdown output of a "regular" `panel: tabset` output (by using the `keep-md: true` option), and then making sure that whatever you produce with `output: asis` matches exactly the div structure that we use. 

`output: asis` is an advanced option, and because we take whatever output produced as the markdown, it's the responsibility of the code using it to generate correct markdown (and yours to produce the correct code).

**Reply by jan-abel-inwt** [reply]
Thank you, the `keep-md: true` option was very helpful.
